File: Verifying an Alien Dictionary.py
""""// Time Complexity : O(n.k); where n is number of words and k is avg length of a word.
// Space Complexity : O(1) as map will at most have 26 chars
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
"""

class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        if len(words) == 1:
            return True
        d = {}

        for i in range(len(order)):
            c = order[i]
            if c not in d:
                d[c] = i

        for i in range(len(words) - 1):
            p1 = 0
            p2 = 0
            while p1 < len(words[i]) and p2 < len(words[i + 1]):
                if d[words[i][p1]] < d[words[i + 1][p2]]:
                    break
                elif d[words[i][p1]] == d[words[i + 1][p2]]:
                    p1 += 1
                    p2 += 1
                else:
                    return False
            if p1 < len(words[i]) and p2 == len(words[i + 1]):
                return False
        return True

File: verifying-an-alien-dictionary.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : -


// Your code here along with comments explaining your approach

class Solution {

    HashMap<Character, Integer> map;

    public boolean isAlienSorted(String[] words, String order) {
        map = new HashMap<>();
        for(int i = 0; i < order.length(); i++){
            char c = order.charAt(i);
            map.put(c, i+1);
        }
        for(int i = 0; i < words.length - 1; i++){
            if(compare(words[i], words[i+1])){
                return false;
            }
        }
        return true;
    }

    private boolean compare(String s1, String s2){
        int m = s1.length(); int n = s2.length();
        for(int i = 0; i < m && i < n; i++){
            char c1 = s1.charAt(i); char c2 = s2.charAt(i);
            if(c1 != c2){
                return map.get(c1) > map.get(c2);
            } 
        }
        return m > n;
    }
}